### PR TITLE
Overhaul handling of Websockets

### PR DESCRIPTION
**Pull request by zv**
This is a new PR without the long list of codeclimate issues. If you want to refer to the old PR, you can [find it here](https://github.com/polyswarm/polyswarmd/pull/257)

# Changes

The biggest change here is in how messages are serialized. The classes are described below, but the highlights are as follows:

1. `FilterManager` registers a filter / formatter tuple (packaged as a `FilterWrapper`)
2. `FilterManager` starts event pool with each `FilterWrapper` (this returns a `gevent.Queue` of results)
3. Each `FilterWrapper` handled by `FilterManager` runs their `spawn_poll_loop(callback)` method, starting up a series of gevents which request new contract filters, format them and place that result on the queue.
4. Back in the Websocket application code, a new result is popped off the `gevent.Queue` mentioned earlier, this result is generally a class of type `WebsocketMessage`, which has the appropriate duner-methods to be passed directly into the underlying websocket `send()` method and broadcast to all users!

## New Classes

As mentioned earlier, this code largely preserves the existing structure of Websocket-Application interaction, but does add some new classes to provide new, clean, features.

### `WebsocketMessage`
All `WebSocket` messages have been extracted into their own class. These classes are mostly small, sharing the majority of their functionality, with only enough information to represent the name that clients like `polyswarm-client` are expecting (`_websocket_event`). Messages which require custom functionality simply override the default behavior.

### `EventLogMessage`
This class provides a method `extract`, intended to convert an `Event` into a structured `dict` by reading the JSONSchema DSL attached to each class (stored in `_extraction_schema`).

The docstring of the conversion function summarizes this DSL:

    It extends jsonschema with several special keys that control extraction from `source`:

        $#fetch - Run a function with args=[source, key, property_schema]
        $#from - Extract this key from `source'

        $#convert - If present and True, format and convert the copied value

    If neither $#fetch or $#from are present, it attempts to fetch the value from `source`
    with the same name as the key.

### `WebsocketFilterMessage`

This class is a combination of the above two classes, which allow it to be used as extraction code or as a message bound for a websocket.

### `FilterManager`

This is a new class which consolidates a mish-mash of functionality already present, such as adding new contract filters and piping their output to a websocket connection (or in this case, representing them as `WebsocketMessages`).

It also has some new features, such as an gevent pool continually filled by the output of fetching new events, automatically creating a new greenlet of itself.

These greenlets also have a local parameter, `wait`, which represents the wait between new calls to request events -- backing off if no data is available or a connection error occurs. 

### `FilterWrapper`

This class presents a unified interface to a single contract event filter and the associated "polling" logic (continuously fetching new contract events). It also provides "formatted" results with an optional `formatter`, which is a piece of code that is run on every result returns, allowing classes like `WebsocketMessage` to receive raw events and pump out formatted messages.

### CODE REVIEW

File: .codeclimate.yml [hunk 1]
```diff
@@ -1,7 +1,17 @@
+version: "2"
+checks:
+  complex-logic:
+    enabled: false
```
Comment (Rizato) [inline .codeclimate.yml]: Idk. I kinda feel these are valuable.

File: src/polyswarmd/websockets/filter.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+            self.flush()
+
+    def setup_event_filters(self, chain):
+        "Setup the most common event filters"
+        if len(self.wrappers) != 0:
+            logger.exception("Attempting to initialize already initialized filter manager")
+            return
+
+        # Setup Latest
+        self.register(chain.w3.eth.filter('latest'), messages.LatestEvent.make(chain.w3.eth))
```
Comment (Rizato) [inline src/polyswarmd/websockets/filter.py]: Backoff should be false here as well, though in an ideal world these should be consistent

File: src/polyswarmd/websockets/serve.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
+import gevent
+
+from geventwebsocket import WebSocketServer, WebSocketApplication, Resource
+
+from .filter import (FilterManager)
+
+
+class EventServer(WebSocketApplication):
```
Comment (Rizato) [inline src/polyswarmd/websockets/serve.py]: Is it possible to deploy this via gunicorn? If so, should we setup a wsgi.py for websockets in addition to serve?

Comment (zv) [reply src/polyswarmd/websockets/serve.py]: I don't know, I've only run it independently (awhile ago). I am currently just keeping this in for future attempts to "isolate" websocket work.

File: .codeclimate.yml (hunk 1 above)
Comment (zv) [reply .codeclimate.yml]: I think they are but I'm no expert in codeclimatology -- How do you instruct it to ignore single files or methods?

**Pull request merged**